digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java

```diff
@@ -65,43 +68,149 @@ private RequireOperator(OperatorContext context, TaskCallbackApi callback)
-                                    attempt.getSessionId(), attempt.getId()));
-                    }
-                    return TaskResult.empty(cf);
-                }
-                else {
-                    // TODO use exponential-backoff to calculate retry interval
-                    throw TaskExecutionException.ofNextPolling(1, ConfigElement.copyOf(request.getLastStateParams()));
-                }
+                throw nextPolling(request.getLastStateParams().deepCopy().set("require_kicked", true));
+            }
+            catch (SessionAttemptConflictException ex) {
+                return processAttempt(ex.getConflictedSession(), lastStateParams, rerunOn, ignoreFailure);
             }
             catch (ResourceNotFoundException ex) {
-                throw new ConfigException(ex);
+                throw new TaskExecutionException(String.format(ENGLISH, "Dependent workflow does not exist. %s, workflowName:%s",
+                        projectIdentifier.transform(ProjectIdentifier::toString).or(""), workflowName));
             }
             catch (ResourceLimitExceededException ex) {
                 throw new TaskExecutionException(ex);
             }
         }
+
+        private TaskResult processAttempt(StoredSessionAttempt attempt, Config lastStateParams, OptionRerunOn rerunOn, boolean ignoreFailure)
+        {
+            if (attempt.getStateFlags().isDone()) {
+                // A flag to distinguish whether the attempt is kicked by require> or previous attempt.
+                boolean requireKicked = lastStateParams.get("require_kicked", boolean.class, false);
+                if (!requireKicked &&  (
+                        rerunOn == OptionRerunOn.ALL ||
+                                rerunOn == OptionRerunOn.FAILED && !attempt.getStateFlags().isSuccess())) {
+
+                    //To force run, set flag gen_retry_attempt_name and do polling
+                    throw nextPolling(lastStateParams.deepCopy().set("rerun_on_retry_attempt_name", UUID.randomUUID().toString()));
+                }
+                if (!ignoreFailure && !attempt.getStateFlags().isSuccess()) {
+                    // ignore_failure is false and the attempt is in error state. Make this operator failed.
+                    throw new TaskExecutionException(String.format(ENGLISH,
+                            "Dependent workflow failed. Session id: %d, attempt id: %d",
+                            attempt.getSessionId(), attempt.getId()));
+                }
+                return TaskResult.empty(cf);
+            }
+            else {
+                // Wait for finish running attempt
+                throw nextPolling(request.getLastStateParams().deepCopy());
+            }
+        }
+
+        private TaskExecutionException nextPolling(Config stateParams)
+        {
+            int iteration = stateParams.get("retry", int.class, 0);
+            stateParams.set("retry", iteration+1);
```

Comment:
Spaces on the both sides of `+` would be better?

Reply:
Thanks! Fixed it.